### PR TITLE
httpbakery: add HTTP checker

### PR DESCRIPTION
**Pull request by rogpeppe**
This will be standard to use in HTTP-based servers.
Also add a single caveat, client-ip-addr and a checker
for that, and a function that can create that from an
existing request.

We also introduce an "error" caveat, which can never be
satisfied, but makes it easy to create caveats from possibly-but-very-rarely-invalid
data without adding lots of error checking. When the caveat
is checked, the error from the check will reflect the orginal
error when the caveat was created.

### CODE REVIEW

File: bakery/checkers/checkers.go [hunk 1]
```diff
@@ -180,3 +192,22 @@ func ParseCaveat(cav string) (cond, arg string, err error) {
...
 	return cav[0:i], cav[i+1:], nil
 }
+
+// ClientIPAddrCaveat returns a caveat that will check whether the
+// client's IP address is as provided.
+// Note that the checkers package provides no specific
+// implementation of the checker for this - that is
+// left to external transport-specific packages.
+func ClientIPAddrCaveat(addr net.IP) bakery.Caveat {
+	if len(addr) != net.IPv4len && len(addr) != net.IPv6len {
```
Comment (mhilton) [inline bakery/checkers/checkers.go:202]: I can foresee a need to provide ranges for this, but perhaps that's a different cidr caveat or something.

File: httpbakery/checkers.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// must be added for each individual set of macaroons
+// that are checked.
+func Checkers(req *http.Request) checkers.Checker {
+	c := httpContext{req}
+	return checkers.Map{
+		checkers.CondClientIPAddr: c.clientIPAddr,
+	}
+}
+
+// hostName implements the host name checker
```
Comment (mhilton) [inline httpbakery/checkers.go]: Wrong function name in doc comment, also wrong comment.

File: httpbakery/checkers.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	if reqIP == nil {
+		return errgo.Newf("cannot parse remote IP address %q", reqHost)
+	}
+	if !reqIP.Equal(ip) {
+		return errgo.Newf("client IP address mismatch, got %s", reqIP)
+	}
+	return nil
+}
+
+// SameRemoteHostCaveat returns a caveat that will check that
```
Comment (mhilton) [inline httpbakery/checkers.go]: Another odd comment.

File: bakery/checkers/checkers.go (hunk 1 above)
Comment (rogpeppe) [inline bakery/checkers/checkers.go:202]: yeah, i think that would be another kind of caveat.

File: httpbakery/checkers.go (hunk 2 above)
Comment (rogpeppe) [inline httpbakery/checkers.go]: fixed

File: httpbakery/checkers.go (hunk 3 above)
Comment (rogpeppe) [inline httpbakery/checkers.go]: fixed

**Comment by alesstimec** [comment]
LGTM with some reservations about ErrorCaveatf.. Please put an example of checking a macaroon for ErrorCaveats into godoc.

File: httpbakery/checkers.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+	"gopkg.in/macaroon-bakery.v0/bakery"
+	"gopkg.in/macaroon-bakery.v0/bakery/checkers"
+)
+
+type httpContext struct {
+	req *http.Request
+}
+
+// HTTP implements the standard HTTP-request checkers.
```
Comment (frankban) [inline httpbakery/checkers.go]: Checkers implements...

File: httpbakery/checkers.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+// that are checked.
+func Checkers(req *http.Request) checkers.Checker {
+	c := httpContext{req}
+	return checkers.Map{
+		checkers.CondClientIPAddr: c.clientIPAddr,
+	}
+}
+
+// clientIPAddr implements the IP client address checker
+// for an http request.
```
Comment (frankban) [inline httpbakery/checkers.go]: // for an HTTP request.
[shrug]

File: httpbakery/checkers.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	return nil
+}
+
+// SameClientIPAddrCaveat returns a caveat that will check that
+// the remote IP address is the same as that in the given HTTP request.
+func SameClientIPAddrCaveat(req *http.Request) bakery.Caveat {
+	if req.RemoteAddr == "" {
+		return checkers.ErrorCaveatf("client has no remote IP address")
+	}
+	reqHost, _, err := net.SplitHostPort(req.RemoteAddr)
```
Comment (frankban) [inline httpbakery/checkers.go:62]: Three checks here are shared between this helper and the caveat checker above.
Is it worth implementing an external function?

File: httpbakery/checkers_test.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	}},
+}, {
+	about: "IPv4 host name declared",
+	checker: checkers.New(httpbakery.Checkers(&http.Request{
+		RemoteAddr: "127.0.0.1:1234",
+	})),
+	checks: []checkTest{{
+		caveat: checkers.ClientIPAddrCaveat(net.IP{127, 0, 0, 1}).Condition,
+	}, {
+		caveat: checkers.ClientIPAddrCaveat(net.IP{127, 0, 0, 1}.To16()).Condition,
```
Comment (frankban) [inline httpbakery/checkers_test.go:52]: Nice!

**Comment by frankban** [comment]
:+1: very nice thank you!

File: httpbakery/checkers.go (hunk 6 above)
Comment (rogpeppe) [inline httpbakery/checkers.go:62]: good catch, done.

File: httpbakery/checkers.go (hunk 5 above)
Comment (rogpeppe) [inline httpbakery/checkers.go]: done

File: httpbakery/checkers.go (hunk 4 above)
Comment (rogpeppe) [inline httpbakery/checkers.go]: done

**Pull request merged**